flocker/node/agents/test/test_blockdevice.py

```diff
@@ -995,6 +1043,222 @@ class UnusableAPI(object):
+        def test_interface(self):
+            """
+            The ``ICalculater`` implemention actually implements the interface.
+            """
+            verifyObject(ICalculater, calculater_factory())
+
+        @given(
+            discovered_datasets=builds(
+                dataset_map_from_iterable,
+                lists(DISCOVERED_DATASET_STRATEGY),
+            ),
+            desired_datasets=builds(
+                dataset_map_from_iterable,
+                lists(DESIRED_DATASET_STRATEGY),
+            ),
+        )
+        def test_returns_changes(self, discovered_datasets, desired_datasets):
+            """
+            ``ICalculater.calculate_changes_for_datasets`` returns a
+            ``IStateChange``.
+            """
+            calculater = calculater_factory()
+            changes = calculater.calculate_changes_for_datasets(
+                discovered_datasets=discovered_datasets,
+                desired_datasets=desired_datasets)
+            self.assertTrue(IStateChange.providedBy(changes))
+
+    return ICalculaterTests
+
+
+class BlockDeviceCalculaterInterfaceTests(
+    make_icalculater_tests(BlockDeviceCalculater)
+):
+    """
+    Tests for ``BlockDeviceCalculater``'s implementation of ``ICalculater``.
+    """
+
+
+def compare_dataset_state(discovered_dataset, desired_dataset):
+    """
+    Compare a discovered dataset to a desired dataset to determine if they have
+    converged.
+
+    .. note:: This ignores ``maximum_size`` as we don't support resizing yet.
+
+    :return: ``bool`` indicating if the datasets correspond.
+    """
+    if discovered_dataset is None:
+        return (desired_dataset is None
+                or desired_dataset.state == DatasetStates.DELETED)
+    if desired_dataset is None:
+        return discovered_dataset.state == DatasetStates.NON_MANIFEST
+    if discovered_dataset.state != desired_dataset.state:
+        return False
+    if discovered_dataset.state == DatasetStates.MOUNTED:
+        return discovered_dataset.mount_point == desired_dataset.mount_point
+    elif discovered_dataset.state == DatasetStates.NON_MANIFEST:
+        return True
+    elif discovered_dataset.state == DatasetStates.DELETED:
+        return True
+    else:
+        raise ValueError("Not possible")
+
+
+def compare_dataset_states(discovered_datasets, desired_datasets):
+    """
+    Compare discovered and desired state of datasets to determine if they have
+    converged.
+
+    .. note:: This ignores ``maximum_size`` as we don't support resizing yet.
+
+    :return: ``bool`` indicating if the datasets correspond.
+    """
+    for dataset_id in set(discovered_datasets) | set(desired_datasets):
+        desired_dataset = desired_datasets.get(dataset_id)
+        discovered_dataset = discovered_datasets.get(dataset_id)
+        result = compare_dataset_state(
+            discovered_dataset=discovered_dataset,
+            desired_dataset=desired_dataset,
+        )
+        if result is False:
```

Comment:
Prefer `if not result:`

See [PEP 8](http://legacy.python.org/dev/peps/pep-0008/).